File: c/KMPAlgorithm.c
#include <stdio.h>
#include <string.h>

void computeLPSArray(char* pat, int M, int* lps)
{
  int len = 0;
  int i;

  for(i = 1; i < M; i++) {
    if(pat[i] == pat[len]) {
      len++;
      lps[i] = len;
    }
    else {
      if(len!=0) {
        len = lps[len - 1];
        i--;
      }
      else {
        lps[i] = len;
      }
    }
  }
}

void KMPSearch(char* pat, char* txt)
{
  int M = strlen(pat);
  int N = strlen(txt);

  int lps[M];
  computeLPSArray(pat, M, lps);

  int i = 0; 
  int j = 0; 
  while(j < N) {
    if(pat[j] == txt[i]) {
      j++;
      i++;
    }
    if(j == M) {
      printf("found pattern at index %d \n", i - j);
      j = lps[j - 1];
    }
    else if (i < N && pat[j] != txt[i]) {
      if(j != 0)
        j = lps[j - 1];
      else
        i = i + 1;
    }
  }
}
int main()
{
  char txt[] = "ABABDABACDABABCABAB";
  char pat[] = "ABABCABAB";
  KMPSearch(pat, txt);
  return 0;
}

File: java/KMPAlgorithm.java
public class KMPAlgorithm
{
  public static int[] computeLPSArray(String pat)
  {
    int M = pat.length();
    int lps[] = new int[M];
    int len = 0;
    int i = 1;
    lps[0] = 0;

    while(i < M)
      {
        if(pat.charAt(i) == pat.charAt(len))
        {
          len++;
          lps[i] = len;
          i++;
        }
        else
        {
          if(len != 0) {
            len = lps[len - 1];
          }
          else
          {
            lps[i] = len;
            i++;
          }
        }
      }
    return lps;
  }

  public static void KMPSearch(String pat, String txt)
  {
    int M = pat.length();
    int N = txt.length();
    int lps[] = computeLPSArray(pat);

    int i = 0;
    int j = 0;
    while (i < N)
      {
        if(pat.charAt(j) == txt.charAt(i)) {
          j++;
          i++;
        }
        if(j == M) 
        {
          System.out.println("found pattern at index" + (i - j));
          j = lps[j - 1];
        }
        else if (i < N && pat.charAt(j) != txt.charAt(i))
          if(j!=0)
            j = lps[j - 1]
          else
            i = i - 1
              }
  }

  public static void main(String args[])
  { 
    String txt = "ABABDABACDABACABA";
    String pat = "ABABCABAB";
    KMPSearch(pat, txt);
  }
  



